pkg/eestream/ranger/reader.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package ranger
+
+import (
+	"bytes"
+	"io"
+)
+
+// A Ranger is a flexible data stream type that allows for more effective
+// pipelining during seeking. A Ranger can return multiple parallel Readers for
+// any subranges.
+type Ranger interface {
+	Size() int64
+	Range(offset, length int64) io.Reader
+}
+
+// FatalReader returns a Reader that always fails with err.
+func FatalReader(err error) io.Reader {
+	return &fatalReader{Err: err}
+}
+
+type fatalReader struct {
+	Err error
+}
+
+func (f *fatalReader) Read(p []byte) (n int, err error) {
+	return 0, f.Err
+}
+
+// ByteRanger turns a byte slice into a Ranger
+type ByteRanger []byte
+
+func (b ByteRanger) Size() int64 { return int64(len(b)) }
```

Comment:
What should be the return value in case of error? `0`, `-1`, something else? Or maybe the function should return `(int64, error)` tuple?

Reply:
my expectation was that Size() would be determined in Ranger constructors (ByteRanger is special and doesn't have a constructor), and Ranger constructors is where the error would happen. If Size() was literally just a cached value accessor (which it probably should be based on how often it's accessed in other utilities) there would never be an error in this method.